src/bsoncxx/document/value.hpp

```diff
@@ -76,6 +81,20 @@ class BSONCXX_API value {
     value(value&&) noexcept = default;
     value& operator=(value&&) noexcept = default;
 
+    // Serializer functions
+    template <typename T,
+              typename std::enable_if<!std::is_same<T, typename array::view>::value, int>::type = 0>
+    explicit value(const T& user_object)
+        : _data(new std::uint8_t[static_cast<std::size_t>(5)],
+                [](std::uint8_t* ptr) { delete[] ptr; }) {
+        to_bson(user_object, *this);
+    }
+    template <typename T>
+    value& operator=(const T& user_object) {
```

Comment:
`.hpp` files are usually just for method declarations. Not definitions. We could move these method definitions to `value.cpp` as is done for the other methods.

Reply:
Separating the declaration and the definition of template functions would cause a linker error, iirc. [Here](https://isocpp.org/wiki/faq/templates#templates-defn-vs-decl) is an example explanation for why this was happening.

Examples like `builder/basic/impl.hpp` and `builder/core.hpp` also adopt this practice of implementing a template function within the header file.

Reply:
Ah yes! Disregard.